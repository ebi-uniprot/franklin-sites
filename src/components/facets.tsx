import { FC, ReactNode, HTMLAttributes, Children } from 'react';
import { Link, useLocation } from 'react-router-dom';
import cn from 'classnames';

import ExpandableList from './expandable-list';

import { formatLargeNumber } from '../utils';

import '../styles/components/facets.scss';

type FacetValue = { label?: ReactNode; value: string; count: number };

export type FacetObject = {
  label?: ReactNode;
  name: string;
  allowMultipleSelection?: boolean;
  values?: FacetValue[];
};

// To hold facets, record of sets
type CustomQueryValue = Record<string, Set<string>>;
// The modified query object, with our custom facet object
export type CustomParsedQuery = Record<string, string | CustomQueryValue>;

/**
 * Takes a search string and parse it, handle facets specifically, keeps them
 * as sets of values
 */
export const parse = (
  string: string,
  queryStringKey = 'facets'
): CustomParsedQuery => {
  const parsed = new URLSearchParams(string);
  const queryStringFacet = parsed.get(queryStringKey);
  const customParsed: CustomParsedQuery = Object.fromEntries(parsed);
  if (!queryStringFacet) {
    return customParsed;
  }
  const facetTokens = (
    Array.isArray(queryStringFacet)
      ? queryStringFacet
      : queryStringFacet.split(',')
  ).map((stringTuple) => stringTuple.split(':'));
  const facets: CustomQueryValue = {};
  for (const [name, value] of facetTokens) {
    if (!facets[name]) {
      facets[name] = new Set();
    }
    facets[name].add(value);
  }
  customParsed[queryStringKey] = facets;
  return customParsed;
};

/**
 * Takes a parsed search object (as generated by the previous "parse" function)
 * and generate a search string
 */
export const stringify = (
  query: CustomParsedQuery,
  queryStringKey = 'facets'
): string => {
  const { [queryStringKey]: facets = {}, ...rest } = query;
  const facetString = Object.entries(facets as CustomQueryValue)
    .map(([name, values]) =>
      Array.from(values).map((value) => `${name}:${value}`)
    )
    .flat()
    .join(',');
  const sp = new URLSearchParams(rest as Record<string, string>);
  if (!facetString) {
    return sp.toString();
  }
  sp.set(queryStringKey, facetString);
  return sp.toString();
};

type FacetProps = {
  /**
   * The facet data to be displayed
   */
  data: FacetObject;
  /**
   * Extra components to be added in the "action" area
   */
  extraActions?: ReactNode;
  /**
   * Key with which to add the facets in the querystring (defaults to "facets")
   */
  queryStringKey?: string;
  /**
   * ClickHandler for specific behaviour
   */
  facetClickHandler?: (event: React.MouseEvent<HTMLElement>) => void;
};

export const Facet: FC<FacetProps & HTMLAttributes<HTMLDivElement>> = ({
  data,
  extraActions,
  queryStringKey = 'facets',
  facetClickHandler,
  ...props
}) => {
  const location = useLocation();
  const search = parse(location.search, queryStringKey);

  if (!data.values?.length) {
    return null;
  }
  return (
    <div {...props}>
      <div className="facet-name">{data.label || data.name}</div>
      <ExpandableList extraActions={extraActions}>
        {data.values.map(({ value, label, count }) => {
          const queryField = search[queryStringKey] as
            | CustomQueryValue
            | undefined;
          const isActive = queryField?.[data.name]?.has(value);

          const facetSet = new Set(
            data.allowMultipleSelection && queryField
              ? queryField[data.name]
              : null
          );
          facetSet[isActive ? 'delete' : 'add'](value);

          const to = {
            ...location,
            search: stringify(
              {
                ...search,
                [queryStringKey]: {
                  ...queryField,
                  [data.name]: facetSet,
                },
              },
              queryStringKey
            ),
          };

          return (
            <Link
              key={`${data.name}_${value}`}
              to={to}
              className={isActive ? 'facet-active' : undefined}
              onClick={facetClickHandler}
            >
              {label || value}
              {` (${formatLargeNumber(count)})`}
            </Link>
          );
        })}
      </ExpandableList>
    </div>
  );
};

type FacetsProps = {
  /**
   * The facet data to be displayed
   */
  data?: FacetObject[];
  /**
   * Extra components to be added in the "action" area, map of <facet name, component>
   */
  extraActionsFor?: Map<string, ReactNode>;
  /**
   * Key with which to add the facets in the querystring (defaults to "facets")
   */
  queryStringKey?: string;
  /**
   * ClickHandler for specific behaviour
   */
  facetClickHandler?: (event: React.MouseEvent<HTMLElement>) => void;
};

export const Facets: FC<FacetsProps & HTMLAttributes<HTMLDivElement>> = ({
  data,
  extraActionsFor,
  queryStringKey = 'facets',
  children,
  className,
  facetClickHandler,
  ...props
}) => {
  if (!(data?.length || Children.count(children))) {
    return null;
  }

  return (
    <div className={cn(className, 'facets')} {...props}>
      <ul className="no-bullet">
        {data?.map((facet) =>
          facet.values?.length ? (
            <li key={facet.name}>
              <Facet
                data={facet}
                extraActions={extraActionsFor?.get(facet.name)}
                queryStringKey={queryStringKey}
                facetClickHandler={facetClickHandler}
              />
            </li>
          ) : null
        )}
        {Children.map(children, (child, index) => {
          if (!child) {
            return null;
          }
          return (
            <li
              key={
                (typeof child === 'object' && 'key' in child && child.key) ||
                index
              }
            >
              {child}
            </li>
          );
        })}
      </ul>
    </div>
  );
};
