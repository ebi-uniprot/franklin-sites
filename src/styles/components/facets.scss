@import '../settings';
@import '../colours';

.facets {
  .facet-name {
    color: $colour-yankees-blue;
    font-weight: 800;
    margin-bottom: $global-margin * 0.25;

    & + ul > li {
      margin-bottom: 1px;
    }
  }

  & > ul > li {
    padding-bottom: $global-margin * 0.5;
    margin-bottom: $global-margin * 0.5;
    border-bottom: 1px solid $colour-platinum;
    position: relative;

    a {
      cursor: pointer;
      display: block;
      width: 100%;
      line-height: $global-lineheight;
      text-align: left;
      padding: $global-margin * 0.3333333333;

      &:hover {
        background: $colour-hover;
      }
    }

    a.facet-active {
      font-weight: bold;
      background: $colour-selected;

      &:after {
        position: absolute;
        right: 1ch;
        content: 'âœ–';
      }

      &:hover,
      &:focus {
        background: $colour-hover;
      }
    }
  }
}
