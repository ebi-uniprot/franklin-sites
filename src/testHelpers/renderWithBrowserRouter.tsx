import { ReactElement } from 'react';
import { BrowserRouter } from 'react-router-dom';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

// Inspired by https://testing-library.com/docs/example-react-router/

const renderWithBrowserRouter = (ui: ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: BrowserRouter });
};

export default renderWithBrowserRouter;
