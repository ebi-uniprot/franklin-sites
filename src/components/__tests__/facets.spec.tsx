import { screen } from '@testing-library/react';

import { Facet, Facets, stringify, parse, CustomParsedQuery } from '../facets';

import renderWithBrowserRouter from '../../testHelpers/renderWithBrowserRouter';

import facets, { statusFacet, proteinsWithFacet } from '../__mocks__/facets';

describe('<Facet />', () => {
  it('should render facet with few data', () => {
    const { asFragment } = renderWithBrowserRouter(
      <Facet data={statusFacet} />
    );
    expect(
      screen.queryByRole('button', { name: /More items/i })
    ).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render facet with a lot of data', () => {
    const { asFragment } = renderWithBrowserRouter(
      <Facet data={proteinsWithFacet} />
    );
    expect(
      screen.getByRole('button', { name: /More items/i })
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<Facets />', () => {
  it('should render multiple facets', () => {
    const { asFragment } = renderWithBrowserRouter(<Facets data={facets} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('<Facet />', () => {
  it('should render facet with few data', () => {
    const { asFragment } = renderWithBrowserRouter(
      <Facet data={statusFacet} />
    );
    expect(
      screen.queryByRole('button', { name: /More items/i })
    ).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render facet with a lot of data', () => {
    const { asFragment } = renderWithBrowserRouter(
      <Facet data={proteinsWithFacet} />
    );
    expect(
      screen.getByRole('button', { name: /More items/i })
    ).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});

const testCases: {
  description: string;
  parsed: CustomParsedQuery;
  query: string;
  key?: string;
}[] = [
  {
    description: 'should handle just one facets',
    parsed: { facets: { identity: new Set(['1.0']) } },
    query: 'facets=identity%3A1.0',
  },
  {
    description: 'should handle just two facets',
    parsed: {
      facets: { blabla: new Set(['value']), identity: new Set(['1.0']) },
    },
    query: 'facets=blabla%3Avalue%2Cidentity%3A1.0',
  },
  {
    description: 'should handle facets and something else',
    parsed: {
      facets: { identity: new Set(['1.0']) },
      query: 'glucose',
    },
    query: 'query=glucose&facets=identity%3A1.0',
  },
  {
    description: 'should handle facets with another name',
    parsed: {
      filter: { identity: new Set(['1.0']) },
      query: 'glucose',
    },
    key: 'filter',
    query: 'query=glucose&filter=identity%3A1.0',
  },
];

describe('parse facets', () => {
  test.each(testCases)('parse $description', ({ parsed, query, key }) => {
    expect(parse(query, key)).toEqual(parsed);
  });
});

describe('stringify facets', () => {
  test.each(testCases)('stringify $description', ({ parsed, query, key }) => {
    expect(stringify(parsed, key)).toEqual(query);
  });
});
