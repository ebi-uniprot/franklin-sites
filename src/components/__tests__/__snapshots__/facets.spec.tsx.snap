// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Facet /> should render facet with a lot of data 1`] = `
<DocumentFragment>
  <div>
    <div
      class="facet-name"
    >
      Proteins with
    </div>
    <ul
      class="expandable-list no-bullet"
    >
      <li>
        <a
          href="/?facets=proteins_with%3A3dstructure"
        >
          3D structure (1,858)
        </a>
      </li>
      <li>
        <a
          href="/?facets=proteins_with%3Aact_site"
        >
          Active site (386,659)
        </a>
      </li>
      <li>
        <a
          href="/?facets=proteins_with%3Aactivity_regulation"
        >
          Activity regulation (30,848)
        </a>
      </li>
      <li>
        <a
          href="/?facets=proteins_with%3Aallergen"
        >
          Allergen (21)
        </a>
      </li>
      <li>
        <a
          href="/?facets=proteins_with%3Aalternative_products"
        >
          Alternative products(isoforms) (1,066)
        </a>
      </li>
      <li>
        <button
          class="button tertiary expandable-list__action"
          data-testid="expandable-message"
          type="button"
        >
          More items
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<Facet /> should render facet with a lot of data 2`] = `
<DocumentFragment>
  <div>
    <div
      class="facet-name"
    >
      Proteins with
    </div>
    <ul
      class="expandable-list no-bullet"
    >
      <li>
        <a
          href="/?facets=proteins_with%3A3dstructure"
        >
          3D structure (1,858)
        </a>
      </li>
      <li>
        <a
          href="/?facets=proteins_with%3Aact_site"
        >
          Active site (386,659)
        </a>
      </li>
      <li>
        <a
          href="/?facets=proteins_with%3Aactivity_regulation"
        >
          Activity regulation (30,848)
        </a>
      </li>
      <li>
        <a
          href="/?facets=proteins_with%3Aallergen"
        >
          Allergen (21)
        </a>
      </li>
      <li>
        <a
          href="/?facets=proteins_with%3Aalternative_products"
        >
          Alternative products(isoforms) (1,066)
        </a>
      </li>
      <li>
        <button
          class="button tertiary expandable-list__action"
          data-testid="expandable-message"
          type="button"
        >
          More items
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<Facet /> should render facet with few data 1`] = `
<DocumentFragment>
  <div>
    <div
      class="facet-name"
    >
      Status
    </div>
    <ul
      class="expandable-list no-bullet"
    >
      <li>
        <a
          href="/?facets=reviewed%3Afalse"
        >
          Unreviewed (TrEMBL) (1,271,454)
        </a>
      </li>
      <li>
        <a
          href="/?facets=reviewed%3Atrue"
        >
          Reviewed (Swiss-Prot) (11,729)
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<Facet /> should render facet with few data 2`] = `
<DocumentFragment>
  <div>
    <div
      class="facet-name"
    >
      Status
    </div>
    <ul
      class="expandable-list no-bullet"
    >
      <li>
        <a
          href="/?facets=reviewed%3Afalse"
        >
          Unreviewed (TrEMBL) (1,271,454)
        </a>
      </li>
      <li>
        <a
          href="/?facets=reviewed%3Atrue"
        >
          Reviewed (Swiss-Prot) (11,729)
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`<Facets /> should render multiple facets 1`] = `
<DocumentFragment>
  <div
    class="facets"
  >
    <ul
      class="no-bullet"
    >
      <li>
        <div>
          <div
            class="facet-name"
          >
            Status
          </div>
          <ul
            class="expandable-list no-bullet"
          >
            <li>
              <a
                href="/?facets=reviewed%3Afalse"
              >
                Unreviewed (TrEMBL) (1,271,454)
              </a>
            </li>
            <li>
              <a
                href="/?facets=reviewed%3Atrue"
              >
                Reviewed (Swiss-Prot) (11,729)
              </a>
            </li>
          </ul>
        </div>
      </li>
      <li>
        <div>
          <div
            class="facet-name"
          >
            Model organisms
          </div>
          <ul
            class="expandable-list no-bullet"
          >
            <li>
              <a
                href="/?facets=model_organism%3AHuman"
              >
                Human (1,470)
              </a>
            </li>
            <li>
              <a
                href="/?facets=model_organism%3AMouse"
              >
                Mouse (1,261)
              </a>
            </li>
            <li>
              <a
                href="/?facets=model_organism%3AA.+thaliana"
              >
                A. thaliana (1,239)
              </a>
            </li>
            <li>
              <a
                href="/?facets=model_organism%3ARat"
              >
                Rat (1,030)
              </a>
            </li>
            <li>
              <a
                href="/?facets=model_organism%3ARice"
              >
                Rice (646)
              </a>
            </li>
          </ul>
        </div>
      </li>
      <li>
        <div>
          <div
            class="facet-name"
          >
            Proteins with
          </div>
          <ul
            class="expandable-list no-bullet"
          >
            <li>
              <a
                href="/?facets=proteins_with%3A3dstructure"
              >
                3D structure (1,858)
              </a>
            </li>
            <li>
              <a
                href="/?facets=proteins_with%3Aact_site"
              >
                Active site (386,659)
              </a>
            </li>
            <li>
              <a
                href="/?facets=proteins_with%3Aactivity_regulation"
              >
                Activity regulation (30,848)
              </a>
            </li>
            <li>
              <a
                href="/?facets=proteins_with%3Aallergen"
              >
                Allergen (21)
              </a>
            </li>
            <li>
              <a
                href="/?facets=proteins_with%3Aalternative_products"
              >
                Alternative products(isoforms) (1,066)
              </a>
            </li>
            <li>
              <button
                class="button tertiary expandable-list__action"
                data-testid="expandable-message"
                type="button"
              >
                More items
              </button>
            </li>
          </ul>
        </div>
      </li>
      <li>
        <div>
          <div
            class="facet-name"
          >
            Protein Existence
          </div>
          <ul
            class="expandable-list no-bullet"
          >
            <li>
              <a
                href="/?facets=existence%3AHOMOLOGY"
              >
                HOMOLOGY (1,000,254)
              </a>
            </li>
            <li>
              <a
                href="/?facets=existence%3APREDICTED"
              >
                PREDICTED (261,161)
              </a>
            </li>
            <li>
              <a
                href="/?facets=existence%3ATRANSCRIPT_LEVEL"
              >
                TRANSCRIPT_LEVEL (15,049)
              </a>
            </li>
            <li>
              <a
                href="/?facets=existence%3APROTEIN_LEVEL"
              >
                PROTEIN_LEVEL (6,700)
              </a>
            </li>
            <li>
              <a
                href="/?facets=existence%3AUNCERTAIN"
              >
                UNCERTAIN (19)
              </a>
            </li>
          </ul>
        </div>
      </li>
      <li>
        <div>
          <div
            class="facet-name"
          >
            Annotation Score
          </div>
          <ul
            class="expandable-list no-bullet"
          >
            <li>
              <a
                href="/?facets=annotation_score%3A2"
              >
                2 (683,263)
              </a>
            </li>
            <li>
              <a
                href="/?facets=annotation_score%3A1"
              >
                1 (377,015)
              </a>
            </li>
            <li>
              <a
                href="/?facets=annotation_score%3A3"
              >
                3 (175,410)
              </a>
            </li>
            <li>
              <a
                href="/?facets=annotation_score%3A4"
              >
                4 (39,819)
              </a>
            </li>
            <li>
              <a
                href="/?facets=annotation_score%3A5"
              >
                5 (7,676)
              </a>
            </li>
          </ul>
        </div>
      </li>
      <li>
        <div>
          <div
            class="facet-name"
          >
            Sequence length
          </div>
          <ul
            class="expandable-list no-bullet"
          >
            <li>
              <a
                href="/?facets=length%3A%5B1+TO+200%5D"
              >
                1 - 200 (101,062)
              </a>
            </li>
            <li>
              <a
                href="/?facets=length%3A%5B201+TO+400%5D"
              >
                201 - 400 (553,257)
              </a>
            </li>
            <li>
              <a
                href="/?facets=length%3A%5B401+TO+600%5D"
              >
                401 - 600 (423,158)
              </a>
            </li>
            <li>
              <a
                href="/?facets=length%3A%5B601+TO+800%5D"
              >
                601 - 800 (120,501)
              </a>
            </li>
            <li>
              <a
                href="/?facets=length%3A%5B801+TO+*%5D"
              >
                &gt;= 801 (85,205)
              </a>
            </li>
          </ul>
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
